ocaml/xapi/repository.ml

```diff
@@ -0,0 +1,975 @@
+  let name = Db.Repository.get_name_label ~__context ~self:ref in
+  let updates_info = parse_updateinfo ~prefix ~name ~hash in
+  let updates_of_hosts, ids_of_updates = List.fold_left (fun (acc1, acc2) host ->
+      let updates =
+        http_get_host_updates_in_json ~__context ~host ~installed:true
+        |> Yojson.Basic.Util.member "updates"
+        |> Yojson.Basic.Util.to_list
+        |> List.map update_of_json
+      in
+      let rpms, uids = List.fold_left (fun (acc_rpms, acc_uids) u ->
+          ( ((Printf.sprintf "%s-%s.%s.rpm" u.name u.new_ver_rel u.arch) :: acc_rpms),
+            (UpdateIdSet.add u.update_id acc_uids) )
+        ) ([], UpdateIdSet.empty) updates
+      in
+      let rec_guidances = List.map (fun g -> `String g)
+          (eval_guidances ~updates_info ~updates ~kind:Recommended)
+      in
+      let abs_guidances = List.map (fun g -> `String g)
+          (eval_guidances ~updates_info ~updates ~kind:Absolute)
+      in
+      let json_of_host = `Assoc [
+          ("ref", `String (Ref.string_of host));
+          ("recommended-guidance", `List rec_guidances);
+          ("absolute-guidance", `List abs_guidances);
+          ("RPMS", `List (List.map (fun r -> `String r) rpms));
+          ("updates", `List (List.map (fun uid -> `String uid) (UpdateIdSet.elements uids)))]
+      in
+      ( (json_of_host :: acc1), (UpdateIdSet.union uids acc2) )
+    ) ([], UpdateIdSet.empty) hosts
+  in
+  `Assoc [
+    ("hosts", `List updates_of_hosts);
+    ("updates", `List (UpdateIdSet.elements ids_of_updates |> List.map (fun x -> `String x)));
+    ("hash", `String hash)]
+
+let get_repository_handler (req : Http.Request.t) s _ =
+  let open Http in
+  let open Xapi_stdext_std.Xstringext in
+  debug "Repository.get_repository_handler URL %s" req.Request.uri ;
+  req.Request.close <- true ;
+  match is_local_pool_repo_enabled () with
+  | true ->
+    (try
+      let repo_name =
+        let l = List.filter_map (fun n ->
+            let path = !Xapi_globs.local_pool_repo_dir ^ "/" ^ n in
+            let stat = Unix.lstat path in
+            begin match stat.Unix.st_kind = Unix.S_DIR with
+              | true -> Some n
+              | false -> None
+            end
+          ) (Array.to_list (Sys.readdir !Xapi_globs.local_pool_repo_dir))
+        in
+        begin match l with
+          | [d] -> d
+          | _ ->
+            let msg = "Multiple or no repositories are under local pool repository directory" in
+            raise Api_errors.(Server_error (internal_error, [msg]))
+        end
+      in
+      let resolved_path =
+        let len = String.length Constants.get_repository_uri in
+        begin match String.sub_to_end req.Request.uri len with
+          | untrusted_path ->
+            untrusted_path
+            |> Filename.concat repo_name
+            |> Filename.concat !Xapi_globs.local_pool_repo_dir
+            |> Uri.pct_decode
+            |> Xapi_stdext_unix.Unixext.resolve_dot_and_dotdot
```

Comment:
It would be good to abstract this "resolve a path safely relative to a root" into a function of its own.
There is already nearly identical code to this elsewhere in xapi, see de2c5887ae4532a9790cda904f78bb0672a16e41 for a security issue we had to fix in that. The code here looks ok, but if we duplicate this kind of security-checking code we risk them getting out-of-date with eachother (e.g. a security fix for one might miss the other) or introducing subtle bugs.

Reply:
Probably best to wrap Fileserver.response_file and have a whitelist of dirs that it is allowed to use.